cxx4opencl/address_spaces.txt

```diff
@@ -0,0 +1,235 @@
+address space will not be compiled unless the address space is explicitly
+specified using address space qualifiers on member functions (see
+<<addrspace-member-function-qualifiers, _Member function qualifier_>>)
+as the conversion between `__constant` and generic is disallowed.
+Member function qualifiers can also be used in case conversion to the
+generic address space is undesirable (even if it is legal). For example,
+a method can be implemented to exploit memory access coalescing for
+segments with memory bank. This not only applies to regular member
+functions but to constructors and destructors too.
+
+[[addrspace-member-function-qualifiers]]
+==== Member function qualifier
+
+{cpp} for OpenCL allows specifying an address space qualifier on member
+functions to signal that they are to be used with objects constructed
+in some specific address space. This works just the same as qualifying
+member functions with `const` or any other qualifiers. The overloading
+resolution will select the candidate with the most specific address
+space if multiple candidates are provided. If there is no conversion
+to an address space among candidates, compilation will fail with a
+diagnostic.
+
+[source,c]
+----------
+struct C {
+   void foo() __local;
+   void foo();
+};
+
+__kernel void bar() {
+  __local C c1;
+  C c2;
+  __constant C c3;
+  c1.foo(); // will resolve to the first foo
+  c2.foo(); // will resolve to the second foo
+  c3.foo(); // error due to mismatching address spaces - can't convert to
+            // __local or generic
+}
+----------
+
+==== Implicit special members
+
+All implicit special members (default, copy, or move constructor, copy or
+move assignment, destructor) will be generated with the generic address
+space.
+
+[source,c]
+----------
+class C {
+  // Has the following implicit definition
+  // void C() /*__generic*/;
+  // void C(const /*__generic*/ C &) /*__generic*/;
+  // void C(/*__generic*/ C &&) /*__generic*/;
+  // operator= '/*__generic*/ C &(/*__generic*/ C &&)'
+  // operator= '/*__generic*/ C &(const /*__generic*/ C &) /*__generic*/
+}
+----------
+
+==== Builtin operators
+
+All builtin operators are available in the specific address spaces, thus
+no conversion to generic address space is performed.
+
+==== Templates
+
+There is no deduction of address spaces in non-pointer/non-reference
+template parameters and dependent types (see <<addrspace-deduction,
+_Deduction_>>). The address space of a template parameter is deduced
+during type deduction if it is not explicitly provided in the
+instantiation.
+
+[source,c]
+----------
+ 1 template<typename T>
+ 2 void foo(T* i){
+ 3   T var;
+ 4 }
+ 5
+ 6 __global int g;
+ 7 void bar(){
+ 8   foo(&g); // error: template instantiation failed as function scope variable
+ 9            // appears to be declared in __global address space (see line 3)
+10 }
+----------
+
+It is not legal to specify multiple different address spaces between
```

Comment:
"multiple different" or would "multiple" be enough too?

Reply:
we accept them if they are multiple but the same :)